### ISSUE TITLE
Latest compcert fails to build

### ISSUE THREAD

**Issue by didickman** [AbsInt/CompCert#151]
On OpenBSD/i386, I get the following error at build time:

```
COQC x86/Op.v
<W> Grammar extension: in [tactic:simple_tactic] some rule has been masked
<W> Grammar extension: in [tactic:simple_tactic] some rule has been masked
<W> Grammar extension: in [tactic:simple_tactic] some rule has been masked
<W> Grammar extension: in [tactic:simple_tactic] some rule has been masked
<W> Grammar extension: in [tactic:simple_tactic] some rule has been masked
File "./x86/Op.v", line 195, characters 0-8:
Anomaly: Uncaught exception Invalid_argument("String.create"). Please report.
```

I'm using these dependencies:
* ocaml-menhir-20161115
* coq-8.5pl3

**Comment by bschommer** [comment]
I use the same configurations and my build works. How did you compile coq, with camlp4 or camlp5 and which ocaml version did you use?

**Comment by didickman** [comment]
Note OpenBSD/i386 is a 32-bit platform, not a 64-bit platform. I understand you might be using a 64-bit platform on your end for the build?

Here are some more versions I'm using:
* ocaml-4.03.0
* ocaml-camlp4-4.03+1

coq is configured with "-usecamlp4". Here's the OpenBSD ports Makefile for coq 8.4pl6 -- pretty similar to what I have locally for 8.5pl3.
https://github.com/openbsd/ports/blob/master/math/coq/Makefile

I'd be happy to provide remote access to the machine if that could help in any way.

**Comment by xavierleroy** [comment]
When run on a 32-bit platform, Coq is known to run into limitations on memory block sizes (inherited from the OCaml runtime system) when compiling definitions to the bytecode of its virtual machine.  I'm surprised that the problem occurs in this particular source file.  Will try to reproduce and debug.

**Comment by xavierleroy** [comment]
Still working on it, but the likely culprit is `decide equality` that would generate bytecode of size O(N^3) where N is the number of constructors of the datatype.  With the addition of 64-bit support to the x86 platform, N (the number of RTL x86 operations) increased a lot recently...  

There are alternatives to `decide equality` but they all need some more work.  I'll look into them later this month.

**Comment by xavierleroy** [comment]
Fixed in commit 647cc8d. The current trunk compiles with a 32-bit Coq and OCaml.  The time it takes to proofcheck x86/Op.v increased significantly, however.
